test/Npgsql.Tests/ConnectionTests.cs

```diff
@@ -1321,6 +1321,29 @@ public async Task TcpKeepalive()
                 Thread.Sleep(Timeout.Infinite);
         }
 
+        [Test, IssueLink("https://github.com/npgsql/npgsql/issues/3511")]
+        public async Task Keepalive_with_failed_transaction()
+        {
+            if (IsMultiplexing)
+                return;
+
+            var csb = new NpgsqlConnectionStringBuilder(ConnectionString)
+            {
+                KeepAlive = 1
+            };
+            using var conn = await OpenConnectionAsync(csb);
+            using var tx = await conn.BeginTransactionAsync();
+
+            Assert.Throws<PostgresException>(() => conn.ExecuteScalar("SELECT non_existent_table"));
+            // Connection is now in a failed transaction state. Wait a bit to allow for the keepalive to execute.
+            Thread.Sleep(3000);
```

Comment:
#3510 adds a method for `PgServerMock`, which waits for a query (actually, it just waits for anything but it doesn't really matter).

Reply:
Or, we can merge it as is, and I fix it on my end.

Reply:
There would still be a race condition though, right? I mean, that method would allow us to know when the server received the keepalive (Sync), but we have no way of knowing when the client received and processed the RFQ... It would probably allow us to reduce the 3 seconds, but do you think it's worth it?

If the mock can't provide a fully determinstic solution for a scenario (without random waiting), we may as well just not use it, no?

Reply:
As a possible solution, we can just send another query (and because of the lock in the StartUserAction, it all should work), but it does sound like a hack...